submission/races.go

```diff
@@ -0,0 +1,284 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package races
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"strings"
+	"sync"
+	"time"
+
+	ct "github.com/google/certificate-transparency-go"
+	"github.com/google/certificate-transparency-go/ctpolicy"
+)
+
+const (
+	// PostBatchInterval is duration between parallel batch call and subsequent requests to Logs within group.
+	// TODO(Mercurrent): optimize to avoid excessive requests.
+	PostBatchInterval = time.Second
+)
+
+// submissionResult holds outcome of a single-log submission.
+type submissionResult struct {
+	sct *ct.SignedCertificateTimestamp
+	err error
+}
+
+type groupState struct {
+	Name    string
+	Success bool
+}
+
+// safeSubmissionState is a submission state-machine for set of Log-groups. When some group is complete cancels all requests that are not needed by any group.
+type safeSubmissionState struct {
+	logToGroups map[string]map[string]bool
+	groupNeeds  map[string]int
+
+	results map[string]*submissionResult
+	cancels map[string]context.CancelFunc
```

Comment:
All these maps mean that safeSubmission group contains a lot of logic that can only be tested as one lump. Possibly, if you made some of them structs + logic then the code would be easier to reason about and the map manipulation would be encapsulated for easier tests.

Reply:
I'm afraid this struct is result of encapsulation of logic.